File: DailyProblem/lowest_missing.py
"""
This problem was asked by Stripe.

Given an array of integers, find the first missing positive integer in linear time and constant space. In other words, find the lowest positive integer that does not exist in the array. The array can contain duplicates and negative numbers as well.

For example, the input [3, 4, -1, 1] should give 2. The input [1, 2, 0] should give 3.

You can modify the input array in-place.
"""

# [3, 4, -1, 1] -> mul -12, sum 7, mul_pos 12
# [1, 2, 0] -> mul 0, sum 3, mul_pos 2


def positive_number_generator(x=1):
    while True:
        yield x
        x += 1


def non_constant_space(nums):
    num_set = set(nums)
    pos_nums = positive_number_generator()
    for p in pos_nums:
        if p not in num_set:
            return p


def solution(nums):

    pass


t1 = [3, 4, -1, 1]
t2 = [1, 2, 0]
functions = [non_constant_space]
tests = [t1, t2]

for i in functions:
    print("\n", i.__name__)
    for j in tests:
        print(i(j))

File: DailyProblem/JS/lowest_missing.js
import "./tester"

function* pos_num_generator(x=1){
    while(true){
        yield x;
        x += 1;
    }
}

function non_constant_space(nums){
    let num_set = new Set(nums);
    let pos_nums = pos_num_generator();
    while(true){
        let n = pos_nums.next().value;
        if(!(num_set.has(n)))
            return n
    }
}

function non_linear_time(nums){
    let pos_nums = pos_num_generator();
    let flag;
    while (true){
        let cur = pos_nums.next().value;
        flag = true;
        for(let i = 0; i < nums.length; i++){
            if(cur === nums[i]) {
                flag = false;
                break;
            }
        }
        if (flag)
            return cur
    }
}

function test_lowest_missing(){
    let t1 = [3, 4, -1, 1];
    let t2 = [1, 2, 0];
    let functions = [non_constant_space, non_linear_time];
    let tests = [t1, t2];

    test.test(tests, functions)
}